File: MVPJava/app/src/main/java/com/vascome/fogtail/api/ApiModule.java
package com.vascome.fogtail.api;

import android.support.annotation.NonNull;

import com.google.gson.Gson;
import com.vascome.fogtail.BuildConfig;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by vasilypopov on 11/22/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

@Module
public class ApiModule {

    @NonNull
    private String baseUrl;

    public ApiModule(@NonNull  String baseUrl) {
        this.baseUrl = baseUrl;
    }

    @Provides @NonNull @Singleton
    public FogtailRestApi provideRestApi(@NonNull OkHttpClient okHttpClient, @NonNull Gson gson) {
        return new Retrofit.Builder()
                .baseUrl(baseUrl)
                .client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create(gson))
                //.addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .validateEagerly(BuildConfig.DEBUG)  // Fail early: check Retrofit configuration at creation time in Debug build.
                .build()
                .create(FogtailRestApi.class);
    }
}

File: MVVMKotlinArchComponent/app/src/main/java/com/vascome/fogtail/di/appmodules/ApiModule.kt
package com.vascome.fogtail.di.appmodules

import com.squareup.moshi.Moshi
import com.vascome.fogtail.BuildConfig
import com.vascome.fogtail.data.api.ApiConfiguration
import com.vascome.fogtail.data.api.FogtailRestApi
import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.moshi.MoshiConverterFactory
import javax.inject.Singleton

/**
 * Created by vasilypopov on 11/22/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

@Module
class ApiModule {

    @Provides
    @Singleton
    fun provideRestApi(okHttpClient:OkHttpClient,
                       moshi:Moshi,
                       config:ApiConfiguration):FogtailRestApi {
        return Retrofit.Builder()
                .baseUrl(config.baseApiUrl)
                .client(okHttpClient)
                .addConverterFactory(MoshiConverterFactory.create(moshi))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .validateEagerly(BuildConfig.DEBUG)  // Fail early: check Retrofit configuration at creation time in Debug build.
                .build()
                .create<FogtailRestApi>(FogtailRestApi::class.java)
    }

    @Provides
    fun provideConfiguration():ApiConfiguration {

        return object : ApiConfiguration {
            override val baseApiUrl: String
                get() = "https://raw.githubusercontent.com/vascome/fogtail/master/MVPJava/app/src/main/assets/"
        }
    }
}